core/src/main/scala/chisel3/experimental/hierarchy/Instantiate.scala

```diff
@@ -106,35 +108,89 @@ object Instantiate {
   }
 
   import chisel3.internal.BuilderContextCache
+  // Include type of module in key since different modules could have the same arguments
   private case class CacheKey[A <: BaseModule](args: Any, tt: ru.WeakTypeTag[A])
       extends BuilderContextCache.Key[Definition[A]]
 
-  /** This is not part of the public API, do not call directly! */
-  def _impl[K, A <: BaseModule: ru.WeakTypeTag](
+  def _instance[K, A <: BaseModule: ru.WeakTypeTag](
     args: K,
     f:    K => A
   )(
     implicit sourceInfo: SourceInfo
-  ): Instance[A] = {
-    val tag = implicitly[ru.WeakTypeTag[A]]
-    // Include type of module in key since different modules could have the same arguments
-    val key = CacheKey(boxAllData(args), tag)
-    val defn = Builder.contextCache
+  ): Instance[A] = Instance.do_apply(_defination(args, f))(sourceInfo)
+
+  /** This is not part of the public API, do not call directly! */
+  def _defination[K, A <: BaseModule: ru.WeakTypeTag](
+    args: K,
+    f:    K => A
+  ): Definition[A] = {
+    Builder.contextCache
       .getOrElseUpdate(
-        key, {
+        CacheKey(boxAllData(args), implicitly[ru.WeakTypeTag[A]]), {
           // The definition needs to have no source locator because otherwise it will be unstably
           // derived from the first invocation of Instantiate for the particular Module
           Definition.do_apply(f(args))(UnlocatableSourceInfo)
         }
       )
       .asInstanceOf[Definition[A]]
-    Instance(defn)
   }
 
   private object internal {
-    // impl cannot be private, but it can be inside of a private object which hides it from the public
+
+    def instance[A <: BaseModule: c.WeakTypeTag](c: Context)(con: c.Tree): c.Tree = {
```

Comment:
I have no idea on how to dedup these macro, since it transforms args... seek help from @jackkoenig